indexer/sql/src/non_blocking_indexer.rs

```diff
@@ -0,0 +1,343 @@
+        models.block.insert(db).await?;
+        entity::transactions::Entity::insert_many(models.transactions)
+            .exec(db)
+            .await?;
+        entity::messages::Entity::insert_many(models.messages)
+            .exec(db)
+            .await?;
+        entity::events::Entity::insert_many(models.events)
+            .exec(db)
+            .await?;
+        Ok(())
+    }
+}
+
+impl NonBlockingIndexer {
+    fn find_or_create<F, R>(&self, block: &BlockInfo, action: F) -> error::Result<R>
+    where
+        F: FnOnce(&mut BlockToIndex) -> error::Result<R>,
+    {
+        let mut blocks = self.blocks.lock().expect("Can't lock blocks");
+        let block_to_index = blocks.entry(block.height).or_insert(BlockToIndex {
+            block_info: *block,
+            txs: vec![],
+        });
+        action(block_to_index)
+    }
+
+    fn find_or_fail(&self, block_height: &u64) -> error::Result<BlockToIndex> {
+        let blocks = self.blocks.lock().expect("Can't lock blocks");
+
+        let block_to_index = match blocks.get(block_height) {
+            Some(block_to_index) => block_to_index,
+            None => {
+                bail!("Block {} not found", block_height);
+            },
+        };
+        Ok(block_to_index.clone())
+    }
+
+    fn remove_or_fail(
+        blocks: Arc<Mutex<HashMap<u64, BlockToIndex>>>,
+        block_height: &u64,
+    ) -> error::Result<BlockToIndex> {
+        let mut blocks = blocks.lock().expect("Can't lock blocks");
+        let block_to_index = match blocks.remove_entry(block_height) {
+            Some(block) => block,
+            None => {
+                bail!("Block {} not found", block_height);
+            },
+        };
+        #[cfg(feature = "tracing")]
+        tracing::debug!(
+            block_height = block_height,
+            blocks_len = blocks.len(),
+            "remove_or_fail called"
+        );
+        Ok(block_to_index.1)
+    }
+}
+
+impl Indexer for NonBlockingIndexer {
+    type Error = crate::error::IndexerError;
+
+    fn start(&mut self) -> error::Result<()> {
+        let context = self.context.clone();
+        block_call(self.runtime.as_ref(), &self.handle, async move {
+            context.migrate_db().await
+        })?;
+
+        self.indexing = true;
+        Ok(())
+    }
+
+    fn shutdown(&mut self) -> error::Result<()> {
+        // Avoid running this twice when called manually and from `Drop`
+        if !self.indexing {
+            return Ok(());
+        }
+        self.indexing = false;
+
+        // NOTE: This is to allow the indexer to commit all db transactions since this is done
+        // async. It just loops quickly making sure no indexing blocks are remaining.
+        for _ in 0..10 {
+            let blocks = self.blocks.lock().expect("Can't lock blocks");
+            if blocks.is_empty() {
+                break;
+            }
+
+            drop(blocks);
+
+            sleep(Duration::from_millis(10));
+        }
```

Comment:
Is possible to to optimize and don't use drop.


```suggestion
        for _ in 0..10 {
            if self.blocks.lock().expect("Can't lock blocks").is_empty() {
                break;
            }
            
            sleep(Duration::from_millis(10));
        }
```